controllers/asosecret_controller.go

```diff
@@ -0,0 +1,290 @@
+type ASOSecretReconciler struct {
+	client.Client
+	Recorder         record.EventRecorder
+	ReconcileTimeout time.Duration
+	WatchFilterValue string
+}
+
+// SetupWithManager initializes this controller with a manager.
+func (asos *ASOSecretReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	_, log, done := tele.StartSpanWithLogger(ctx,
+		"controllers.ASOSecretReconciler.SetupWithManager",
+		tele.KVP("controller", "ASOSecret"),
+	)
+	defer done()
+
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		WithOptions(options).
+		For(&infrav1.AzureCluster{}).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(log, asos.WatchFilterValue)).
+		WithEventFilter(predicates.ResourceIsNotExternallyManaged(log)).
+		Named("ASOSecret").
+		Build(asos)
+	if err != nil {
+		return errors.Wrap(err, "error creating controller")
+	}
+
+	// Add a watch on infrav1.AzureManagedControlPlane if Cluster API 'MachinePool' feature is enabled.
+	if err = c.Watch(
+		&source.Kind{Type: &infrav1.AzureManagedControlPlane{}},
+		&handler.EnqueueRequestForObject{},
+		predicates.ResourceNotPausedAndHasFilterLabel(log, asos.WatchFilterValue),
+	); err != nil {
+		return errors.Wrap(err, "failed adding a watch for ready clusters")
+	}
+
+	// Add a watch on clusterv1.Cluster object for unpause notifications.
+	if err = c.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
+		handler.EnqueueRequestsFromMapFunc(util.ClusterToInfrastructureMapFunc(ctx, infrav1.GroupVersion.WithKind("AzureCluster"), mgr.GetClient(), &infrav1.AzureCluster{})),
+		predicates.ClusterUnpaused(log),
+		predicates.ResourceNotPausedAndHasFilterLabel(log, asos.WatchFilterValue),
+	); err != nil {
+		return errors.Wrap(err, "failed adding a watch for ready clusters")
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the ASO secrets associated with AzureCluster objects.
+func (asos *ASOSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(asos.ReconcileTimeout))
+	defer cancel()
+
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "controllers.ASOSecret.Reconcile",
+		tele.KVP("namespace", req.Namespace),
+		tele.KVP("name", req.Name),
+		tele.KVP("kind", "AzureCluster"),
+	)
+	defer done()
+
+	log = log.WithValues("namespace", req.Namespace, "AzureCluster", req.Name)
+
+	// asoSecretOwner is the resource that created the identity. This could be either an AzureCluster or AzureManagedControlPlane (if AKS is enabled).
+	// check for AzureCluster first and if it is not found, check for AzureManagedControlPlane.
+	var asoSecretOwner client.Object
+
+	azureCluster := &infrav1.AzureCluster{}
+	// Fetch the AzureCluster or AzureManagedControlPlane instance
+	asoSecretOwner = azureCluster
+	err := asos.Get(ctx, req.NamespacedName, azureCluster)
+	if err != nil && apierrors.IsNotFound(err) {
```

Comment:
Nit: could probably reorganize the ifs for consistency with the other controllers as follows

```
if err != nil {
	if err apierrors.IsNotFound(err) {
		...
	}

	return reconcile.Result{}, err
}
```

Reply:
Tried to reorganize them a bit, but they still look a bit heavy unfortunately :(